src/threads/synch.c

```diff
@@ -114,10 +114,16 @@ sema_up (struct semaphore *sema)
 
   old_level = intr_disable ();
   if (!list_empty (&sema->waiters)) 
-    thread_unblock (list_entry (list_pop_front (&sema->waiters),
-                                struct thread, elem));
+  {
+    // sort list to support case that priorities could be changed on blocked threads
+    list_sort(&sema->waiters, thread_priority_higher, NULL);
+    thread_unblock (list_entry (list_pop_front (&sema->waiters), struct thread, elem));
+  }
   sema->value++;
   intr_set_level (old_level);
+  // there is a chance that unblocked thread has higher priority,
+  // let it to be rescheduled yelding CPU on the current thread (test priority-sema)
+  thread_yield ();
```

Comment:
potentially new `preempt` function can be used agains unblocked thread?

Reply:
Yes, good catch!

I tried to do it, but there is a problem with current `preempt` implementation. It does not work here for some test and I did not investigate why :(
Probably because we can not call it without disabling interrupts..

If you have time, you could help to investigate what is happening here ;)